### DISCUSSION TITLE
getting started questions from a new user

### DISCUSSION THREAD

**Discussion by yuribudilov** [sreeise/graph-rs-sdk#426] [Q&A] [answered]
Hello everyone, thank you for this Rust Microsoft Graph SDK !!

I am not entirely new to Rust (but I am new to async Rust and tokio and I am new to web development in general).
I am also new to dealing with Microsoft Identity Security and Microsoft Graph with Microsoft Office365.

So, I am trying to learn it all by developing my own Desktop GUI App (I am not using any browser).
I am planning to use Rust GUI software such as **egui crate** or any other Rust GUI that will work with this SDK.
I want to sign-in from Windows PC or from Linux Ubuntu PC and them to access my own personal email from hotmail/outlook. 

I am looking for any tips on how to sign-in to Microsoft Office365 email from Windows 10/11 or Ubuntu laptops app and how to get various token authorizations to be able to manipulate my own email (read, delete, send, etc) the way I can do using Windows Outlook rich UI app or Edge Web Browser.

Any pointers to what I need to read and which parts of this Rust SDK example code I should use to go through Security would be greatly appreciated. There are many examples on how to sign in, but I dont know which ones apply to my use case. 
The access tokens are hard-coded into static strings in examples (github) which makes it harder for me to put all elements together from api to sign-on to get id/token to authorization/access token api. 
I am happy to RTFM if you could provide me with some pointers for my use case (use device flow perhaps?).
I don't aim to become Microsoft Identity guru, I just want to develop code to access things correctly. 
There are so many materials on this topic that it seems I will be reading for weeks just to figure out which set of API and code flows I need to use. 
For languages like C# and JavaScript the Microsoft language SDK seems to cover up the complexity and end-to-end code samples exist.
In Rust they are hard to find, hence my question here.

thank you!

**Comment by sreeise** [comment]
Hey @yuribudilov 

Ok first this:

> The access tokens are hard-coded into static strings in examples (github) which makes it harder for me to put all elements together from api to sign-on to get id/token to authorization/access token api.

Those are just examples. The sdk doesn't require a `static &str`. You can use just a regular `&str` in place of the static ones anywhere you see them in the examples.

Thats a lot of different areas that your asking about. Let me see if I answer from what I understand after reading it.

> So, I am trying to learn it all by developing my own Desktop GUI App (I am not using any browser).
> I am planning to use Rust GUI software such as egui crate or any other Rust GUI that will work with this SDK.
> I want to sign-in from Windows PC or from Linux Ubuntu PC and them to access my own personal email from hotmail/outlook.

Are you saying your wanting to add sign in from the GUI you will be creating? If so, thats not something ive really looked into or attempted to use the oauth from this crate for. Although I am not saying its not possible.

I divided this reply into two sections so its easier to understand. The first section below explains my advice to you and what I think you could consider doing based on what you have told me. The second section gets into the more difficult areas of what your trying to do with more details.

## The easier way

> I am not entirely new to Rust (but I am new to async Rust and tokio and I am new to web development in general).
I am also new to dealing with Microsoft Identity Security and Microsoft Graph with Microsoft Office365.

Thats great your into Rust. And web development is a fun field. What your wanting to do, from what I understand, is absoutely possible. But with that said, what your wanting to do may not be the easist if your new to web development. You will have to read a lot and you will learn a lot. If you willing to, and since your new to web development, I suggest just starting out by seeing if you can just do the authorization code flow, and copy the authorization code from your browser by hand. That would be this example: https://github.com/sreeise/graph-rs-sdk/blob/master/examples/oauth/auth_code_grant.rs (requires application setup in Azure first).

I would highly suggest that you start simple. Take the time to read the authorization flows and maybe read some on desktop apps authorization flows https://learn.microsoft.com/en-us/azure/active-directory/develop/scenario-desktop-overview

With a personal microsoft account, which is what your account is if im not mistaken (unless its a school or work account), there are some flows you may think will work like the device code flow or resource owner passoword credentials, however, those typically do not work for personal Microsoft accounts. You can look at the individual authorization flows here: https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow

The very first thing you will want to do, if you havnt, is create an an application in Azure active directory. Lastly, you could try adding your personal microsoft account to the active directory and that may allow you to use other authorization flows such as device code flow or resource owner password credential. The resource owner password credential is not usually reccomended though due to security concerns but if its just your account its probably fine. But check up on those to see if they allow personal Microsoft accounts just to make sure im not mistaken.

## The more difficult way

>I am looking for any tips on how to sign-in to Microsoft Office365 email from Windows 10/11 or Ubuntu laptops app and how to get various token authorizations to be able to manipulate my own email (read, delete, send, etc) the way I can do using Windows Outlook rich UI app or Edge Web Browser.

So it sounds like you are wanting to create a desktop application? If not let me know. There are specific authorization flows for desktop applications. Normally desktop applications on Windows use Windows sdks to add sign on. There is a Windows sdk for Rust and if thats your goal then you can ask the maintainers of that crate if its possible to do sign on and get an access token via a GUI. The crate for the [Windows sdk is here](https://crates.io/crates/windows). The GitHub for that crate is [windows-rs](https://github.com/microsoft/windows-rs/tree/master). 

You will still need to create an application in Azure active directory I am assuming so maybe its possible that you can do that.
Have you already done the steps necessary to create an application

But if your really wanting to build this from scratch, the authorization flows for native desktop apps typically work, in a production scenario (meaning many users are using it and its developed and maintained properly), by having a server side web API that handles the authorization flow and that is communicating with the desktop app and passing that required info to the desktop app or native app. In that case the web API would use the authorization code flow which means your listening for a redirect, getting the access code, and then passing that access code to the native app. 

There are ways to do it without having a web API, such as authentication brokers but thats not possible currently in this sdk, see https://learn.microsoft.com/en-us/azure/active-directory/develop/scenario-desktop-acquire-token-wam for more info on that. 


> I am happy to RTFM if you could provide me with some pointers for my use case (use device flow perhaps?). 

I might only be 29 but I don't pay attention anymore just like old people so I don't know what RTFM is. Sorry haha.

Device code flow, if im not mistaken, does not work for personal microsoft accounts.

Theres also the integrated windows authentication, also not in the SDK, and also only works with domain joined accounts. Unfortuantely WAM might just be the best way to do it but again, its not currently in the SDK and I have not done enough research in order to implement it in this SDK as of yet or to know whether its possible.

Lastly, just in general yes its a lot to read, but if you actually wanting to learn OAuth, which is really a subsection of web development, things like these are what you have to do. So I reccommend: https://learn.microsoft.com/en-us/azure/active-directory/develop/authentication-vs-authorization

**Reply by yuribudilov** [reply]
Many thanks for your reply! Easy answers to your questions first: I do have Azure AD User (my personal Azure dev account) and I have also added my Application Authorization to Azure AD as per MS documentation. The Rust application I am trying to create is something similar to MS Outlook GUI App present on Windows-10 to deal with personal emails. This app will read my email, allow me to list my messages, to delete messages, send messages, etc. In order to do this my Rust GUI app will need to authenticate user (i.e. me) and then to obtain necessary access tokens to call various Microsoft Graph APIs to deal with my email. So, I am **not** trying to create a web-browser app at all but just an ordinary Rust GUI standalone app. I want to build a rich-GUI stand-alone Rust app running on Windows and Linux/Ubuntu which authenticates via Microsoft Identity and then calls Microsoft Graph APIs to communicate with Microsoft Email system. Initially this app will only run on my Windows-10/11 laptop and the next step is also to run it on my Linux Ubuntu laptop as well (and may be on my MacOS too, optionally?). The Rust GUI crate that I am thinking of using is called **egui** (it is meant to be portable Rust GUI, it runs on Windows, Linux and MacOS). With this in mind, I am trying to figure out which specific security/identity Microsoft services I can use from your graph-rs-sdk to be able to achieve this task. (PS: RTFM is a rude 4-letter acronym, stands for _Read The F***ing Manual_ this acronym was very widely used when I first entered IT industry in late 1980s!! **LOL** ). You have provided many examples here which is great, I am trying to see which of those I need to use for my Rust app. THANK YOU very much!! Rust Rocks!

**Reply by sreeise** [reply]
Well.. now I feel bad. Not only am I too old to pay attention now but im also too young to know what that meant 😆

**Reply by sreeise** [reply]
I didn't even know this until I went backed and looked at MSAL.NET, Microsofts authorization library in .net, but apparently they are somehow able to see the browser URL and get the access code in that library. Makes me wonder how it can be done. Thats something I will need to look into. https://learn.microsoft.com/en-us/azure/active-directory/develop/msal-net-web-browsers

But anyway, feel free to ask more questions and I will do my best to answer. If you are planning on using non-personal microsoft accounts than yeah its posibble to use other authorization flows. I am doing some work on the oauth part of this sdk and it will look similar to using the .net ones but the available features will obviously be less than what they can offer. Thanks for letting me know that this was helpful.

**Reply by sreeise** [reply]
I went back and read the device code flow one again, you actually can use that for personal Microsoft accounts. I apologize about that. 

https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code#device-authorization-request

Its basically requiring the user to enter a code into another device or a browser to authorize them if your ok with just showing the user the code and having them enter it similar to logging in on a tv to an account or on a mobile device.

I don't have an example for that one though. The code flow example in the examples/oauth directory is not device code flow.

You can see what each type of app, desktop or web or mobile, typically uses for authorization here: https://learn.microsoft.com/en-us/azure/active-directory/develop/authentication-flows-app-scenarios#scenarios-and-supported-authentication-flows

**Reply by sreeise** [reply]
One last thing. If you use something like a webview than this gets a lot easier. You might have to use separate UI libraries for it but its probably worth it to get a good sign on experience. I just searched for webview at crates.io and the most downloaded is https://crates.io/crates/wry

There are some webview crates that look to be a wrapper around the Windows webview in .net which does have specific library functionality for getting signed in users authorization. Up to you just thought I would mention it.

Edit: This is often controversial to do though because of security concerns meaning using webviews for authorization. Meant to mention that earlier. See https://www.oauth.com/oauth2-servers/mobile-and-native-apps/security-considerations/

**Reply by yuribudilov** [reply]
thank you very much! Lots of reading, as I feared. My latest thoughts are to use REST APIs directly from Rust (via hyper?) so mimic what CURL does. It is crude and perhaps too complex but if CURL can do it using GET and POST as shown here: [ https://learn.microsoft.com/en-us/graph/auth-v2-service ] then Rust using hyper and serde crates should be able to? What do you think? From what I saw, MMAL.Net does not support Rust, C and C++, based on the this link: [ https://learn.microsoft.com/en-us/azure/active-directory/develop/authentication-flows-app-scenarios#scenarios-and-supported-authentication-flows ]

**Reply by sreeise** [reply, marked as answer]
The first link you mentioned, https://learn.microsoft.com/en-us/graph/auth-v2-service, is just the client credentials grant flow which is what this project uses for tests and there is an example of it here: https://github.com/sreeise/graph-rs-sdk/blob/master/examples/oauth/client_credentials.rs

Yeah they don't support Rust for sure. I don't know about C or C++ but I don't think they support those either. 

> thank you very much! Lots of reading, as I feared. My latest thoughts are to use REST APIs directly from Rust (via hyper?) so mimic what CURL does. 

You can try doing that but thats pretty much what this crate is already doing so if there is an authorization flow, such as client credentials, that this crate offers than it will probably be easier than writing your own. 

I also added an example for the device code flow and updated the sdk in order to be able to do that flow in version 1.1.1. That example is here: https://github.com/sreeise/graph-rs-sdk/blob/master/examples/oauth/device_code.rs

**Reply by yuribudilov** [reply]
thank you very much, I have just constructed the very crude CURL Rest API prototype, just from Windows 10 command shell and it seems to work but it's very early days. I will keep on digging, and much obliged for your support. Keep on doing the good job with Rust! Using Azure with Rust seems to be much more work than AWS and Google Clouds. The AWS has a proper Rust SDK now (official SDK, in beta) and I have used Rust with GCP too as well as with AWS. It appears that Microsoft cloud is well behind and relying on kindness of strangers to support Rust in Azure. Disappointing given the Rust benefits over other main-stream languages are now well known and documented.

**Comment by sreeise** [comment]
@yuribudilov 

Also this:

> For languages like C# and JavaScript the Microsoft language SDK seems to cover up the complexity and end-to-end code samples exist.
In Rust they are hard to find, hence my question here.

Yeah Microsoft is a very large multinational corporation that has the resources to hire thousands of developers and create all that content for you to look at but they don't currently support their own Rust sdk which is why ive continued to maintain this repo. But yeah I can't make up for the fact that im not a multinational corporation with billions in revenue. Im joking with you by the way. I just laughed though when I read what you said that you must think I have infinite resources. Your all good. Good chuckle.
